doc/rationale.md

```diff
@@ -0,0 +1,84 @@
+# Configy rationale
+
+The following explains the rationale versus various decisions made in Configy.
+
+## Use cases, Internal vs External schema
+
+Configy was originally designed with an application-first mindset: a developer should
+design their config file based on a `struct` definition they chose, and would potentially
+have a slightly different configuration file depending on the problem domain (for example,
+the `@Key("name")` attribute).
+
+While this is still the primary use-case for Configy, it is also very convenient to
+write a validator for schema that are outside of the application's control,
+such as `dub.json`, `.gitlab-ci.yml` / GitHub actions file, or a replacement for
+a JSON schema.
+
+The first use case (when the schema is under the developer's control) is referred to
+as an "Internal schema", while the second use case (when the schema is outside of the
+developer's control) is referred to as an "External schema". While both are supported,
+and Configy aims to have sane default that are likely to appear in external schemas,
+it might sometimes be required to "drop down" to using hooks (`fromString`, `fromYAML`)
+to accurately control deserialization for external schema.
```

Comment:
Whether the schema is internal or external seems to be a distraction here.

The facts are:

- Configy primarily supports an _opinionated subset_ of all possible YAML schemas.
- Application authors may want to read schemas outside of this subset, either because they're not under their control ("external") or they just disagree with Configy's opinionated decisions.

Reply:
That's a very good point. Edited accordingly.